src/Types.jl

```diff
@@ -451,293 +452,184 @@ function relative_project_path(ctx::Context, path::String)
-    given_abspath = nothing
+    pkg.special_action = PKGSPEC_REPO_ADDED
+    # The first goal is to populate pkg.repo.source if that wasn't given explicitly
     if pkg.repo.source === nothing
         @assert pkg.repo.rev !== nothing
-        # First, we try resolving against the manifest to avoid updating registries if at all possible.
+        # First, we try resolving against the manifest and current registry to avoid updating registries if at all possible.
         # This also handles the case where we _only_ wish to switch the tracking branch for a package.
         manifest_resolve!(ctx, [pkg]; force=true)
-        # If we could not resolve against manifest, resolve against the registry.
-        if pkg.name === nothing || pkg.uuid === nothing
-            @goto reg_resolve
-        end
-        entry = manifest_info(ctx, pkg.uuid)
-        # We still have to check this case, recall that name and UUID could both be given explicitly.
-        entry !== nothing || @goto reg_resolve 
-        if entry.pinned
-            pkg.tree_hash = entry.tree_hash # TODO why do we need this?
-            return false
-        elseif entry.repo.source !== nothing # reuse source in manifest
-            # TODO check consistency with entry
-            pkg.repo.source = entry.repo.source
-            @goto do_clone
-        end
-        @label reg_resolve
-        update_registries(ctx) # If the information not in manifest, we have no choice but to update the registry.
-        registry_resolve!(ctx, pkg)
-        if pkg.name === nothing || pkg.uuid === nothing
-            pkgerror("Package $(err_rep(pkg)) could not be found in a registry or a manifest.")
+        if isresolved(pkg)
+            entry = manifest_info(ctx, pkg.uuid)
+            if entry !== nothing && entry.repo.source !== nothing # reuse source in manifest
+                pkg.repo.source = entry.repo.source
+            end
         end
-        paths = registered_paths(ctx, pkg.uuid)
-        # We still have to check this case, both name/UUID can be given explicitly.
-        if isempty(paths)
-            pkgerror("Package with UUID `$(pkg.uuid)` could not be found in a registry.")
+        if pkg.repo.source === nothing
+            set_repo_source_from_registry!(ctx, pkg)
         end
-        _, pkg.repo.source = Types.registered_info(ctx, pkg.uuid, "repo")[1]
     end
-    @label do_clone
     @assert pkg.repo.source !== nothing
+
+    # We now have the source of the package repo, check if it is a local path and if that exists
     if !isurl(pkg.repo.source)
-        given_abspath = isabspath(pkg.repo.source)
-        pkg.repo.source = try
-            realpath(pkg.repo.source)
-        catch
+        if isdir(pkg.repo.source)
+            if !isdir(joinpath(pkg.repo.source, ".git"))
+                pkgerror("Did not find a git repository at `$(pkg.repo.source)`")
+            end
+            pkg.repo.source = safe_realpath(pkg.repo.source)
```

Comment:
It seems this does not preserve the relative v absolute path to a local repo. (e.g. as opposed to https://github.com/JuliaLang/Pkg.jl/blob/b29b799c667e6bcf5bb954eec9f05b0a273ddfbe/src/Types.jl#L715-L716)

Reply:
Yep, fixed.